structs-in-rust/article.md

```diff
@@ -0,0 +1,419 @@
+```
+
+The answer is that both `Debug` and `std::fmt::Display` are **traits**, and we need to implement them for `Point` if we want to print out the struct.
+
+### What is a trait?
+
+A trait is similar to an *interface* in Java or a *typeclass* in Haskell. It defines certain functionality that we might expect from some type.
+
+Usually, traits declare a list of methods that can be called on the types that implement this trait.
+
+Here is a list of commonly-used traits from the standard library:
+
+- [`Debug`](https://doc.rust-lang.org/std/fmt/trait.Debug.html). Used to format (and print) a value using `{:?}`.
+- [`Clone`](https://doc.rust-lang.org/std/clone/trait.Clone.html). Used to get a duplicate of a value.
+- [`Default`](https://doc.rust-lang.org/std/default/trait.Default.html). Used to create a default instance of type.
+- [`Eq`](https://doc.rust-lang.org/std/cmp/trait.Eq.html). Used to compare values for equality.
+
+In our example, the `Point` needs an implementation of the `Debug` trait. It is required by the `println!()` macro.
+
+There are two ways to implement a trait.
+
+- Manually write an implementation.
+- **Derive** it. The compiler is capable of providing basic implementations for a fixed list of traits via the `#[derive]` macro.
+
+In this article, we'll cover only the second option because it is the shortest way to achieve our needs. 
+
+Let's add a new line before the `Point` definition:
+
+```
+#[derive(Debug, PartialEq, Eq, Default)]
+struct Point {
+    x: i32,
+    y: i32,
+}
+```
+
+#### Usage of Debug
+```
+let p = Point { x: 0, y: 1 };
+println!("{:?}", p);
+```
+
+Now the compiler is not complaining, let's run our program.
+
+```
+> cargo run
+Point { x: 0, y: 1 }
+```
+
+We didn't write anything related to formatting `Point` ourselves, but we already have a decent-looking output. Neat!
+
+#### Usage of Eq
+
+You may have noticed additional `Eq` and `PartialEq` traits.  Since we derived those traits, we can compare two `Point`s for equality:
+
+```
+let a = Point { x: 25, y: 50 };
+let b = Point { x: 100, y: 100 };
+let c = Point { x: 25, y: 50 };
+
+assert!(a == c);
+assert!(a != b);
+assert!(b != c);
+```
+
+#### Usage of Default
+
+For this example, we need to define another struct.
+
+```
+#[derive(Debug, PartialEq, Eq, Default)]
+struct DoubleBool(bool, bool);
+
+let point: Point = Default::default();
+let double_bool: DoubleBool = Default::default();
+
+assert_eq!(point, Point { x: 0, y: 0 });
+assert_eq!(double_bool, DoubleBool(false, false));
+```
+
+The most important lines in this snippet are those where we create `point` and `double_bool`. Because both `Point` and `DoubleBool` implement `Default`, we can invoke the `default()` method of the `Default` trait in each case to generate an empty value of type `Point` and `DoubleBool` respectively.
+
+*Exercise: try to remove explicit type annotations from lines 3 and 4 and see what the compiler tells you.*
```

Comment:
I'd add the word "Why?" at the end to make it feel more like an exercise.

Reply:
Ok